tests/soak_test/map_soak_test.py

```diff
@@ -0,0 +1,200 @@
+
+import argparse
+import threading
+import logging
+import time
+import random
+
+from hazelcast.client import HazelcastClient
+from hazelcast.serialization.api import IdentifiedDataSerializable
+from hazelcast.predicate import between
+
+THREAD_COUNT = 32
+ENTRY_COUNT = 10000
+STATS_DISPLAY_SECONDS = 10.0
+
+
+class Statistic(object):
+    def __init__(self, thread_id):
+        self.id = thread_id
+        self._total_ops = 0
+        self._lock = threading.Lock()
+
+    def get_total_operation_count_and_reset(self):
+        with self._lock:
+            total = self._total_ops
+            self._total_ops = 0
+            return total
+
+    def increment_operation_count(self):
+        with self._lock:
+            self._total_ops += 1
+
+
+class SimpleEntryProcessor(IdentifiedDataSerializable):
+    CLASS_ID = 1
+    FACTORY_ID = 66
+
+    def __init__(self, value):
+        self.value = value
+
+    def read_data(self, object_data_input):
+        pass
+
+    def write_data(self, object_data_output):
+        object_data_output.write_utf(self.value)
+
+    def get_class_id(self):
+        return self.CLASS_ID
+
+    def get_factory_id(self):
+        return self.FACTORY_ID
+
+
+def listener(_):
+    pass
+
+
+def start():
+    thread_count_before = threading.active_count()
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--hour", default=48.0, type=float, help="Duration of the test in hours")
+    parser.add_argument("--addresses", default="127.0.0.1", type=str,
+                        help="List of cluster member addresses separated by -")
+    parser.add_argument("--log", default="default_log_file", type=str, help="Name of the log file")
+
+    args = parser.parse_args()
+
+    hour_limit = args.hour
+    addresses = args.addresses
+    log_file = args.log
+
+    logging.basicConfig(filename=log_file,
+                        filemode="w",
+                        format="%(asctime)s %(message)s",
+                        datefmt="%H:%M:%S",
+                        level=logging.INFO)
+
+    try:
+        client = HazelcastClient(
+            cluster_members=addresses.split("-"),
+            data_serializable_factories={
+                SimpleEntryProcessor.FACTORY_ID: {
+                    SimpleEntryProcessor.CLASS_ID: SimpleEntryProcessor
+                }
+            }
+        )
+    except Exception:
+        logging.exception("Client failed to start")
+        return
+
+    processor = SimpleEntryProcessor("test")
+    test_map = client.get_map("test-map").blocking()
+    test_map.add_entry_listener(False, added_func=listener, removed_func=listener, updated_func=listener)
+
+    logging.info("Soak test operations are starting!")
+    logging.info("* " * 20 + "\n")
+
+    test_failed = [False]
+
+    def run(stats):
+        end_time = time.time() + hour_limit * 60 * 60
+        while time.time() < end_time:
+            if test_failed[0]:
```

Comment:
Is this check thread safe? Or to be more precise, does Python guarantee that the updated list value will be published to all threads?

Reply:
Well, Python does not have a memory model, so these details are not specified anywhere. 

But, CPython has GIL, so I guess every update should be visible to the thread that will hold the GIL next time. It is mentioned here slightly https://www.python.org/dev/peps/pep-0583/#rationale

Reply:
BTW are we going to be any safer if we used one of built-in collection classes instead? But probably, such details aren't critical for a soak test.

Reply:
Do you mean the funny usage of `test_failed[0]`? If it is, this is related to fact that you cannot edit a variable defined in the encapsulating scope from the closure. (There is `nonlocal` keyword in Python3 for this case, but not in Python2). So, I have to define something that can be modified inside the closure and picked the `list` since it is the simplest alternative.

Reply:
I see. OK, let's leave it this way then.